packages/beacon-node/src/api/impl/validator/index.ts

```diff
@@ -490,196 +513,185 @@ export function getValidatorApi({
+              }) as ReturnType<typeof produceEngineFullBlockOrContents>)
+            : Promise.reject<ReturnType<typeof produceEngineFullBlockOrContents>>(engineDisabledError),
+        ],
+        {
+          resolveTimeoutMs: BLOCK_PRODUCTION_RACE_CUTOFF_MS,
+          raceTimeoutMs: BLOCK_PRODUCTION_RACE_TIMEOUT_MS,
+          signal: controller.signal,
         }
-      } else if (blindedBlockPromise !== null && fullBlockPromise === null) {
-        blindedBlock = await blindedBlockPromise;
-        fullBlock = null;
-      } else if (blindedBlockPromise === null && fullBlockPromise !== null) {
-        blindedBlock = null;
-        fullBlock = await fullBlockPromise;
-      } else {
+      );
+
+      if (
+        builder.status === "rejected" &&
+        builder.reason === builderDisabledError &&
+        engine.status === "rejected" &&
+        engine.reason === engineDisabledError
+      ) {
         throw Error(
           `Internal Error: Neither builder nor execution proposal flow activated isBuilderEnabled=${isBuilderEnabled} builderSelection=${builderSelection}`
         );
       }
 
-      const builderPayloadValue = blindedBlock?.executionPayloadValue ?? BigInt(0);
-      const enginePayloadValue = fullBlock?.executionPayloadValue ?? BigInt(0);
-      const consensusBlockValueBuilder = blindedBlock?.consensusBlockValue ?? BigInt(0);
-      const consensusBlockValueEngine = fullBlock?.consensusBlockValue ?? BigInt(0);
+      if (builder.status === "rejected" && builder.reason !== builderDisabledError) {
+        logger.warn(
+          "Builder failed to produce the block",
+          {
+            durationMs: builder.durationMs,
+            slot,
+          },
+          builder.reason as Error
+        );
+      }
 
-      const blockValueBuilder = builderPayloadValue + consensusBlockValueBuilder; // Total block value is in wei
-      const blockValueEngine = enginePayloadValue + consensusBlockValueEngine; // Total block value is in wei
+      if (engine.status === "rejected" && engine.reason !== engineDisabledError) {
+        logger.warn(
+          "Engine failed to produce the block",
+          {
+            durationMs: engine.durationMs,
+            slot,
+          },
+          engine.reason as Error
+        );
+      }
 
-      let executionPayloadSource: ProducedBlockSource | null = null;
-      const shouldOverrideBuilder = fullBlock?.shouldOverrideBuilder ?? false;
+      if (builder.status === "pending" && engine.status === "pending") {
+        throw Error(
+          `Builder and execution both timeout to proposed the block in ${BLOCK_PRODUCTION_RACE_TIMEOUT_MS}ms`
+        );
+      }
 
-      // handle the builder override case separately
-      if (shouldOverrideBuilder === true) {
-        executionPayloadSource = ProducedBlockSource.engine;
+      // handle builder override case separately
+      const shouldOverrideBuilder = engine.status === "fulfilled" && engine.value.shouldOverrideBuilder;
```

Comment:
Just wanna note here that `PromiseWithStatus` is really unsafe to use, we lose type safety a native Promise would add

e.g. it is possible to call `.value` and `.reason` without checking if promise is rejected / fulfilled already and if it is not we throw a runtime error. This is not possible with native promises as you are forced to use `await` or `.then`, which makes it much safer to use.

It feels a bit overengineered but if it not possible to solve this by solely relying on native primitives like `null` or `Promise` it might be justified.

Reply:
`PromiseWithStatus` is a promise, you can use `await` and `try/catch` with it.

Reply:
but that's not how we use it